File: farandula-java/src/main/java/com/nearsoft/farandula/flightmanagers/amadeus/AmadeusFlightManager.java
package com.nearsoft.farandula.flightmanagers.amadeus;

import com.jayway.jsonpath.JsonPath;
import com.jayway.jsonpath.ReadContext;
import com.nearsoft.farandula.exceptions.ErrorType;
import com.nearsoft.farandula.exceptions.FarandulaException;
import com.nearsoft.farandula.flightmanagers.FlightManager;
import com.nearsoft.farandula.models.*;
import net.minidev.json.JSONArray;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import java.io.IOException;
import java.io.InputStream;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static com.nearsoft.farandula.utilities.CabinClassParser.getCabinClassType;
import static com.nearsoft.farandula.utilities.CurrencyIATACodesHelper.buildPrice;
import static com.nearsoft.farandula.utilities.NestedMapsHelper.getValueOf;

/**
 * Created by pruiz on 4/20/17.
 */
public class AmadeusFlightManager implements FlightManager {

    private final OkHttpClient.Builder _builder = new OkHttpClient.Builder();
    private static String apiKey;
    private static Map<String, String> locationsMap = new HashMap<>();
    private static Map<String,String> airlinesCodeMap = new HashMap<>();

    static {
        Properties props = new Properties();
        try {
            props.load(AmadeusFlightManager.class.getResourceAsStream("/config.properties"));
            fillReferenceMaps();
            apiKey = props.getProperty("amadeus.apikey");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void fillReferenceMaps() throws IOException {
        Properties properties = new Properties();
        properties.load(AmadeusFlightManager.class.getResourceAsStream("/amadeus/locations.properties"));
        for (String key : properties.stringPropertyNames()) {
            String value = properties.getProperty(key);
            locationsMap.put(key, value);
        }

        properties.clear();
        properties.load( AmadeusFlightManager.class.getResourceAsStream("/airlinesCode.properties"));
        for (String key : properties.stringPropertyNames()) {
            String value = properties.getProperty(key);
            airlinesCodeMap.put(key, value);
        }

    }

    private OkHttpClient buildHttpClient() {
        if (_builder.interceptors().isEmpty()) {
            _builder.connectTimeout(1, TimeUnit.MINUTES);
            _builder.readTimeout(1, TimeUnit.MINUTES);
        }
        return _builder.build();
    }

    @Override
    public List<Itinerary> getAvail(SearchCommand search) throws FarandulaException {

        try {
            Request request = buildRequestForAvail(search);
            InputStream responseStream = sendRequest(request);
            return parseAvailResponse(responseStream);

        } catch (Exception e) {
            throw new FarandulaException(e, ErrorType.AVAILABILITY_ERROR, "error retrieving availability");
        }

    }

    public String getTimeZone(String location_code) throws IOException, FarandulaException {

        String url_api = buildLocationURL(location_code);
        Request request = buildRequest(url_api);
        Response response = buildHttpClient().newCall(request).execute();
        InputStream responseStream = response.body().byteStream();
        ReadContext context = JsonPath.parse(responseStream);
        JSONArray airports = context.read("$..airports[*]");

        List<String> timezones = airports
                .stream()
                .filter(airport -> ((Map<String, Object>) airport).get("code").equals(location_code))
                .map(airport -> {
                    return (String) ((Map<String, Object>) airport).get("timezone");
                })
                .collect(Collectors.toCollection(ArrayList::new));

        return timezones.size() > 0 ? timezones.get(0) : "";

    }


    public List<Itinerary> parseAvailResponse(InputStream response) throws IOException {


        List<Itinerary> itineraries = new ArrayList<>();
        ReadContext ctx = JsonPath.parse(response);
        JSONArray results = ctx.read("$..results[*]");

        for ( Object result : results ){

            Itinerary itineraryResult = new Itinerary();

            Map<String, Object> resultMap = (Map<String, Object>) result;

            JSONArray arrayItineraries = (JSONArray) resultMap.get("itineraries");

            for ( Object itinerary : arrayItineraries ){

               buildAirLegs((Map<String, Object>) itinerary, itineraryResult );

            }

            //pricing
            Map<String, Object> fareMap = (Map<String, Object>) ((LinkedHashMap) result).get("fare");
            itineraryResult.setPrice( getPrices( fareMap ) );

            itineraries.add( itineraryResult );

        }

        return itineraries;

    }

    private void buildAirLegs(Map<String, Object> itineraryMap, Itinerary itineraryResult ) {

        //adds departure leg
        Map<String, Object> outbound = (Map<String, Object>) itineraryMap.get("outbound");
        JSONArray outboundFlights = (JSONArray) outbound.get("flights");
        AirLeg departureLeg = getAirleg(outboundFlights);
        itineraryResult.getAirlegs().add( departureLeg );

        //adds returnings leg
        Map<String, Object> inbound = (Map<String, Object>) itineraryMap.get("inbound");
        JSONArray inboundFlights = (JSONArray) inbound.get("flights");
        AirLeg returnigLeg= getAirleg(inboundFlights);
        itineraryResult.getAirlegs().add(returnigLeg );

    }

    private Fares getPrices(Map<String, Object> pricingInfoData) {

        //price
        Fares fares =  new Fares();
        Price totalPrice = buildPrice( getValueOf( pricingInfoData , "total_price", String.class ) );
        fares.setTotalPrice( totalPrice );

        if ( pricingInfoData.get("price_per_adult") != null  ) {
            fares.setPricePerAdult( buildPrice( getValueOf(pricingInfoData, "price_per_adult.total_fare", String.class)));
            fares.setTaxPerAdult( buildPrice( getValueOf(pricingInfoData, "price_per_adult.tax", String.class) ) );
        }

        if ( pricingInfoData.get("price_per_child") != null  ) {
            fares.setPricePerChild( buildPrice(getValueOf(pricingInfoData, "price_per_child.total_fare", String.class)));
            fares.setTaxPerChild( buildPrice(getValueOf(pricingInfoData, "price_per_child.tax", String.class)));
        }

        if ( pricingInfoData.get("price_per_infant") != null  ) {
            fares.setPricePerInfant(buildPrice(getValueOf(pricingInfoData, "price_per_infant.total_fare", String.class)));
            fares.setTaxPerInfant(buildPrice(getValueOf(pricingInfoData, "price_per_infant.tax", String.class)));
        }

        return fares;
    }

    private AirLeg getAirleg(JSONArray outboundFlights) {
        LinkedList<Segment> segmentsOutbound = outboundFlights
                .stream()
                .map(segment -> {
                    try {
                        return buildSegment((Map<String, Object>) segment);
                    } catch (IOException e) {
                        e.printStackTrace();
                    } catch (FarandulaException e) {
                        e.printStackTrace();
                    }
                    return null;
                })
                .collect(Collectors.toCollection(LinkedList::new));

        AirLeg leg = new AirLeg();
        leg.setId("tempID");
        leg.setDepartureAirportCode(segmentsOutbound.get(0).getDepartureAirportCode());
        leg.setDepartingDate(segmentsOutbound.get(0).getDepartureDate());
        leg.setArrivalAirportCode(segmentsOutbound.get(segmentsOutbound.size() - 1).getArrivalAirportCode());
        leg.setArrivalDate(segmentsOutbound.get(segmentsOutbound.size() - 1).getArrivalDate());
        leg.setSegments(segmentsOutbound);
        return leg;
    }

    private Segment buildSegment(Map<String, Object> segmentMap) throws IOException, FarandulaException {

        //departure
        Map<String, Object> departureAirportData = (Map<String, Object>) segmentMap.get("origin");

        //arrival
        Map<String, Object> arrivalAirportData = (Map<String, Object>) segmentMap.get("destination");

        //booking_info
        Map<String, Object> bookingInfoData = (Map<String, Object>) segmentMap.get("booking_info");

        //pricing_info
        Map<String, Object> pricingInfoData = (Map<String, Object>) segmentMap.get("fare");

        //Airleg information
        Segment seg = new Segment();
        seg.setAirlineIconPath("");
        seg.setOperatingAirlineCode((String) segmentMap.get("operating_airline"));
        seg.setOperatingAirlineName( airlinesCodeMap.get( seg.getOperatingAirlineCode() ) );
        seg.setOperatingFlightNumber( (String) segmentMap.get("flight_number")  );

        seg.setMarketingAirlineCode((String) segmentMap.get("marketing_airline"));
        seg.setMarketingAirlineName( airlinesCodeMap.get( seg.getMarketingAirlineCode() ) );
        seg.setMarketingFlightNumber((String) segmentMap.get("flight_number"));

        seg.setAirplaneData((String) segmentMap.get("aircraft"));

        CabinClassType classTravel =  getCabinClassType( (String) bookingInfoData.get("travel_class") );
        int numberOfSeats = (int) bookingInfoData.get("seats_remaining");
        List<Seat> seats = new ArrayList<>();
        for ( int i = 0 ; i < numberOfSeats ; i++ ){
            Seat seat = new Seat();
            //amadeus does not indicate the seat place
            seat.setPlace("");
            seat.setClassCabin( classTravel );
            seats.add( seat );
        }
        seg.setSeatsAvailable(seats);

        //departure stuff
        seg.setDepartureAirportCode((String) departureAirportData.get("airport"));
        seg.setDepartureTerminal( (String) departureAirportData.get("terminal")) ;
        LocalDateTime departureDateTime = LocalDateTime.parse(
                (String) segmentMap.get("departs_at"), DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        seg.setDepartureDate(departureDateTime);

        //arrival stuff
        seg.setArrivalAirportCode((String) arrivalAirportData.get("airport"));
        seg.setArrivalTerminal( (String) arrivalAirportData.get("terminal") );
        LocalDateTime arrivalDateTime = LocalDateTime.parse(
                (String) segmentMap.get("arrives_at"), DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        seg.setArrivalDate(arrivalDateTime);

        String departureTimeZone = "";
        if (locationsMap.containsKey(seg.getDepartureAirportCode())) {
            departureTimeZone = locationsMap.get(seg.getDepartureAirportCode());
        } else {
            departureTimeZone = getTimeZone(seg.getDepartureAirportCode());
            locationsMap.put(seg.getDepartureAirportCode(), departureTimeZone);
        }

        String arrivalTimeZone = "";
        if (locationsMap.containsKey(seg.getArrivalAirportCode())) {
            arrivalTimeZone = locationsMap.get(seg.getArrivalAirportCode());
        } else {
            arrivalTimeZone = getTimeZone(seg.getArrivalAirportCode());
            locationsMap.put(seg.getArrivalAirportCode(), arrivalTimeZone);
        }

        long diffInMinutes = 0;
        if (departureTimeZone.equals(arrivalTimeZone)) {
            diffInMinutes = Duration.between(departureDateTime, arrivalDateTime).toMinutes();
        } else {
            if (departureTimeZone.isEmpty() || arrivalTimeZone.isEmpty()) {
                //todo check this case, what should we do when is impossible to get the zone,
                // example when location is 'xyz'
                diffInMinutes = 0;
            } else {
                ZonedDateTime departureWithZone = departureDateTime.atZone(ZoneId.of(departureTimeZone));
                ZonedDateTime arrivalWithZone = arrivalDateTime.atZone(ZoneId.of(arrivalTimeZone));
                diffInMinutes = Duration.between(departureWithZone, arrivalWithZone).toMinutes();
            }
        }

        seg.setDuration(diffInMinutes);

        return seg;
    }

    public Request buildRequestForAvail(SearchCommand search) {

        String url_api = buildTargetURLFromSearch(search);
        return buildRequest(url_api);
    }

    public Request buildRequest(String url_api) {
        final Request.Builder builder = new Request.Builder();
        builder.url(url_api);
        builder.get();
        return builder.build();
    }

    public String buildTargetURLFromSearch(SearchCommand search) {

        List<String> departureDateList = new ArrayList<>();
        for(  LocalDateTime departing :  search.getDepartingDates() ){
            departureDateList.add( departing.format( DateTimeFormatter.ISO_LOCAL_DATE  ) );
        }



        String apiURL = "https://api.sandbox.amadeus.com/v1.2/flights/low-fare-search?";

        apiURL +="apikey=" + apiKey;

        for ( int i = 0 ; i < search.getArrivalAirports().size() ; i++ ) {
            apiURL += "&origin=" + search.getDepartureAirports().get( i )
                    + "&destination=" + search.getArrivalAirports().get( i )
                    +"&departure_date=" + departureDateList.get(i);
        }

        if ( search.getType() == FlightType.ROUNDTRIP) {
            String arrivalDate = search.getReturningDates().get(0).format( DateTimeFormatter.ISO_LOCAL_DATE  ) ;
            apiURL += "&return_date=" + arrivalDate;
        }


        for ( Map.Entry< PassengerType, List<Passenger> > entry : search.getPassengersMap().entrySet()  ){
            apiURL += "&"+  entry.getKey().toString().toLowerCase() + "=" + entry.getValue().size();
        }

        apiURL += "&number_of_results=" + search.getOffSet();

        //cabins for amadeus.
        switch ( search.getCabinClass() ){

            case ECONOMY:
                apiURL+= "&travel_class=ECONOMY";
                break;

            case PREMIUM_ECONOMY:
                apiURL+= "&travel_class=PREMIUM_ECONOMY";
                break;

            case FIRST:
                apiURL+= "&travel_class=FIRST";
                break;

            case BUSINESS:
                apiURL+= "&travel_class=FIRST";
                break;

            default:
                apiURL+= "&travel_class=ECONOMY";
        }

        return apiURL;
    }

    String buildLocationURL(String location) {

        String api = "http://api.sandbox.amadeus.com/v1.2/location/"
                + location
                + "/?apikey=" + apiKey;
        return api;
    }

    public InputStream sendRequest(Request request) throws IOException, FarandulaException {
        final Response response = buildHttpClient().newCall(request).execute();
        return response.body().byteStream();
    }

}

File: farandula-ruby/lib/farandula/flight_managers/amadeus/amadeus_flight_manager.rb
require 'rest-client'
require 'yaml'
require_relative '../flight_manager.rb'
require 'farandula/models/itinerary'
require 'farandula/models/air_leg'
require 'farandula/models/segment'
require 'farandula/models/seat'
require 'farandula/models/fares'
require 'farandula/models/price'

require_relative '../../constants.rb'
require 'time'
require 'tzinfo'

module Farandula
  module FlightManagers
    module Amadeus

      class AmadeusFlightManager < FlightManager

        attr_accessor :api_key

        def initialize(api_key)
          @api_key = api_key

          @airline_code_map = YAML.load_file(File.dirname(__FILE__) + '/../../assets/amadeus/' + "airlinesCode.yml")

          @locations_map = YAML.load_file(File.dirname(__FILE__) + '/../../assets/amadeus/' + "locations.yml")

        end

        def get_avail(search_form)

          request = Amadeus::Request.new

          url_request = request.build_target_url_from_search!(search_form, api_key)

          response = ""
          url_request.each_with_index {| url, index |
            response = RestClient.get url
          }

          build_itineries(response)

        end

        def build_itineries (response)

          itineries_list = []

          parsed = JSON.parse(response)

          itinerary_id = 0

          parsed["results"].each {|result|

            #pricing
            # Map<String, Object> fareMap = (Map<String, Object>) ((LinkedHashMap) result).get("fare");
            # itineraryResult.setPrice(getPrices(fareMap));

            fares = get_prices( result['fare'])

            result['itineraries'].each {|itinerary|

              itineray_result = Farandula::Itinerary.new

              itineray_result.fares = fares

              itinerary_id +=1

              itineries_list << build_air_legs(itineray_result, itinerary, itinerary_id)

            }
          }

          itineries_list

        end

        private
        def build_air_legs (itinerary_object, itinerary_json_result, id = nil)

          itinerary_object.departure_air_legs = build_leg(itinerary_json_result['outbound']['flights'], 1)

          if itinerary_json_result['inbound']
            itinerary_object.returning_air_legs = build_leg(itinerary_json_result['inbound']['flights'], 1)
          end

          itinerary_object.id = id

          itinerary_object

        end

        def build_leg (jsonObject, id = nil)

          segment_id = 0
          segments_array = jsonObject.map {|segment|
            segment_id =+1
            build_segment(segment, segment_id)
          }

          leg = AirLeg.new
          leg.id = id;
          leg.departure_airport_code = segments_array[0].departure_airport_code
          leg.departure_date = segments_array[0].departure_date
          leg.arrival_airport_code = segments_array[segments_array.size - 1].arrival_airport_code
          leg.arrival_date = segments_array[segments_array.size - 1].arrival_date
          leg.segments = segments_array
          return leg;

        end

        def build_segment(segmentJson, segment_key)

          departure_airport_data = segmentJson['origin']

          arrival_airport_data = segmentJson['destination']

          booking_info_data = segmentJson['booking_info']

          segment = Farandula::Segment.new

          segment.key = segment_key

          segment.operating_airline_code = segmentJson['operating_airline']
          segment.operating_airline_name = @airline_code_map[segment.operating_airline_code]
          segment.operating_flight_number = segmentJson['flight_number']

          #if marketing and operating are the same airline
          if segmentJson['marketing_airline']
            segment.marketing_airline_code = segmentJson['marketing_airline']
          elsif segment.marketing_airline_code = segmentJson['operating_airline']
          end
          segment.marketing_airline_name = @airline_code_map[segment.marketing_airline_code]
          segment.marketing_flight_number = segmentJson['flight_number']

          #airplane data stuff i.e. boeing 771
          segment.airplane_data = segmentJson['aircraft']

          # class travel stuff
          class_travel = get_cabin_class_type(booking_info_data['travel_class'])
          number_of_seat = booking_info_data["seats_remaining"].to_i

          seats = []
          number_of_seat.times do
            seat = Farandula::Seat.new
            seat.cabin = class_travel
            #amadeus does not indicate the seat place
            seat.place = ""
            seats << seat
          end
          segment.seats_available = seats

          #departure stuff
          segment.departure_airport_code = departure_airport_data['airport']
          segment.departure_terminal = departure_airport_data['terminal']
          departure_date_time = format_date(Time.parse (segmentJson['departs_at']))
          segment.departure_date = departure_date_time

          #arrival stuff
          segment.arrival_airport_code = arrival_airport_data['airport']
          segment.arrival_terminal = arrival_airport_data['terminal']
          arrival_date_time = format_date (Time.parse(segmentJson['arrives_at']))
          segment.arrival_date = arrival_date_time


          if @locations_map[segment.departure_airport_code]
            departure_location = @locations_map[segment.departure_airport_code]
          end

          if @locations_map[segment.arrival_airport_code]
            arrival_location = @locations_map[segment.arrival_airport_code]
          end

          # duration flight stuff
          if departure_location.equal? arrival_location
            duration = arrival_date_time - departure_date_time
          elsif tz = TZInfo::Timezone.get(departure_location)
            tz2 = TZInfo::Timezone.get(arrival_location)
            # seconds to minutes
            duration = (tz2.local_to_utc(Time.parse(segmentJson['arrives_at'])) -
                tz.local_to_utc(Time.parse (segmentJson['departs_at']))) /60
          end

          segment.duration = duration
          segment

        end

        def get_cabin_class_type (travel_class_info)

          case travel_class_info.downcase
            when "economy"
              CabinClassType::ECONOMY
            when "business"
              CabinClassType::BUSINESS
            when "premiumeconomy"
              CabinClassType::PREMIUM_ECONOMY
            when "first"
              CabinClassType::FIRST
            when "economy/coach"
              CabinClassType::ECONOMYCOACH
            when "premium_economy"
              CabinClassType::PREMIUM_ECONOMY
            else
              CabinClassType::OTHER
          end

        end

        def get_prices ( fare )

          fares = Farandula::Fares.new

          fares.base_price = build_price( fare['total_price'] )

          if fare['price_per_adult']
            fares.price_per_adult = build_price( fare['price_per_adult']['total_fare'] )
            fares.tax_per_adult   = build_price( fare['price_per_adult']['tax'] )
          end

          if fare['price_per_child']
            fares.price_per_child = build_price( fare['price_per_child']['total_fare'] )
            fares.tax_per_child   = build_price( fare['price_per_child']['tax'] )
          end

          if fare['price_per_infant']
            fares.price_per_child = build_price( fare['price_per_infant']['total_fare'] )
            fares.tax_per_child   = build_price( fare['price_per_infant']['tax'] )
          end

          fares

        end

        def build_price( pricing_info_data  )

          currency = pricing_info_data.gsub( /[^a-zA-Z]+/, '' )
          if  ''.eql? currency
              currency  = "US"
              amount    = pricing_info_data
          else
              amount = pricing_info_data.gsub( currency, '')
          end

          priceResult = Farandula::Price.new
          priceResult.amount  = amount.to_f
          priceResult.currency_code = currency
          priceResult

        end

        def format_date(date)
          date.strftime('%Y-%m-%dT%H:%M:%S.%L%z')
        end

      end #ends amadeusflightmanager
    end #ends Amadeus
  end #ends FlightManagers
end # ends Farandula